onnx/defs/loss/defs.cc

```diff
@@ -0,0 +1,113 @@
+    L, if reduction = 'none'
+)DOC";
+
+bool BuildContextDependentFunctionBodySCE(const FunctionBodyBuildContext& ctx, const OpSchema& schema, FunctionProto& functionProto) {
+  std::vector<FunctionBodyHelper::NodeDef> body;
+  body.push_back({{"X_SM"}, "Softmax", {"scores"}});
+  body.push_back({{"X_Log"}, "Log", {"X_SM"}});
+  if (ctx.hasInput(2)) {
+    body.push_back({{"output"}, "NegativeLogLikelihoodLoss", {"X_Log", "labels"}, {MakeAttribute("reduction", "mean")}});
+  } else {
+    body.push_back({{"output"}, "NegativeLogLikelihoodLoss", {"X_Log", "labels", "weights"}, {MakeAttribute("reduction", "mean")}});
+  }
+
+  auto func_nodes = FunctionBodyHelper::BuildNodes(body);
+  for (const auto node : func_nodes) {
+    auto new_node = functionProto.add_node();
+    new_node->CopyFrom(node);
+  }
+
+  schema.BuildFunction(functionProto);
+  return true;
+}
+
+ONNX_OPERATOR_SET_SCHEMA(
+    SoftmaxCrossEntropyLoss,
+    12,
+    OpSchema()
+        .SetDoc(SoftmaxCrossEntropyLoss_ver12_doc)
+        .Attr(
+            "reduction",
+            reduction_doc,
+            AttributeProto::STRING,
+            std::string("mean"))
+        .Input(
+            0,
+            "scores",
+            "The predicted outputs with shape [batch_size, class_size], or "
+            "[batch_size, class_size, d1, d2 , ..., dk], where K is the number of dimensions.",
+            "T")
+        .Input(
+            1,
+            "labels",
+            "The ground truth output tensor, with shape [batch_size], or "
+            "[batch_size, d1, d2 , ..., dk], where K is the number of dimensions."
+            "Usualy, it's a one-hot representation of groud-truth class.",
+            "T")
+         .Input(
+            2,
+            "weights",
+            "A manual rescaling weight given to each class. If given, it has to "
+            "be a 1D Tensor assigning weight to each of the classes. Otherwise, "
+            "it is treated as if having all ones.",
+            "T",
+            OpSchema::Optional)
+        .Output(
+            0,
+            "output",
+            "Weighted loss float Tensor. If reduction is 'none', this has the "
+            "shape of [batch_size], or [batch_size, d1, d2, ..., dk] in case of "
+            "K-dimensional loss. Otherwise, it is a scalar.",
+            "T")
+        .TypeConstraint(
+            "T",
+            {"tensor(float16)", "tensor(float)", "tensor(double)"},
+            "Constrain input and output types to float tensors.")
+        .SetContextDependentFunctionBodyBuilder(BuildContextDependentFunctionBodySCE)
+        .TypeAndShapeInferenceFunction([](InferenceContext& ctx) {
+            propagateElemTypeFromInputToOutput(ctx, 0, 0);
+            std::string reduction = getAttribute(ctx, "reduction", "mean");
+            if (reduction.compare("none") == 0 && hasInputShape(ctx, 1)) {
+                propagateShapeFromInputToOutput(ctx, 1, 0);
+            }
```

Comment:
if reduction is none, the output is a scalar, right? we can set the output shape for that case also.

Reply:
If reduction is none, output is the same shape as labels. If reduction is not none then output is scalar. Do you want me to add that case?

Reply:
Yes, sorry that's what I meant. Please do add it.